translator/reader/reader.cpp

```diff
@@ -158,25 +159,74 @@ unsigned int UsdArnoldReader::RenderThread(void *data)
+    // element. Thus, every time we'll read a prim, the last element of this 
+    // stack will represent its input primvars that it inherits (see #282)
+    std::vector<std::vector<UsdGeomPrimvar> > &primvarsStack = threadData->context.GetPrimvarsStack();
+    primvarsStack.clear(); 
+    primvarsStack.reserve(64); // reserve first to avoid frequent memory allocations
+    primvarsStack.push_back(std::vector<UsdGeomPrimvar>()); // add an empty element first
+    
     // Traverse the stage, either the full one, or starting from a root primitive
-    // (in case an object_path is set).
-    UsdPrimRange range = (rootPrim) ? UsdPrimRange(*rootPrim) : reader->GetStage()->Traverse();
+    // (in case an object_path is set). We need to have "pre" and "post" visits in order
+    // to keep track of the primvars list at every point in the hierarchy.
+    UsdPrimRange range = UsdPrimRange::PreAndPostVisit((rootPrim) ? 
+                    *rootPrim : reader->GetStage()->GetPseudoRoot());
     for (auto iter = range.begin(); iter != range.end(); ++iter) {
         const UsdPrim &prim(*iter);
-        // Check if that primitive is set as being invisible.
-        // If so, skip it and prune its children to avoid useless conversions
 
-        // Special case for arnold schemas, they don't inherit from UsdGeomImageable
-        // but we author these attributes nevertheless
         std::string objType = prim.GetTypeName().GetText();
-        if (prim.IsA<UsdGeomImageable>() || objType.substr(0, 6) == "Arnold") {
+        // skip untyped primitives
+        if (objType.empty())
+            continue;
+
+        // We traverse every primitive twice : once from root to leaf, 
+        // then back from leaf to root. We don't want to anything during "post" visits
+        // apart from popping the last element in the primvars stack for xforms.
+        // This way, the last element in the stack will always match the current 
+        // set of primvars
+        if (iter.IsPostVisit()) {
+            if (objType == "Xform")
+                primvarsStack.pop_back();
+            continue; 
+        }
+
+        bool checkVisibility = false;
+        if (objType == "Xform") {
+            
+            // Get the inheritable primvars for this xform, by giving its parent ones as input
+            UsdGeomPrimvarsAPI primvarsAPI(prim);
+            std::vector<UsdGeomPrimvar> primvars = 
+                primvarsAPI.FindIncrementallyInheritablePrimvars(primvarsStack.back());
+            
+            // if the returned vector is empty, we want to keep using the same list as our parent
+            if (primvars.empty())
+                primvarsStack.push_back(primvarsStack.back());
+            else
+                primvarsStack.push_back(primvars); // primvars were modified for this xform
+
+            checkVisibility = true;
+        } else {
+            // Special case for arnold schemas, they don't inherit from UsdGeomImageable
```

Comment:
Is this still required even after #558 ?

Reply:
when we'll release the schemas, I guess we'll be able to remove this hack. But for now, I don't think USD can be aware of this inheritance